File: week4/16236/16236.cpp
#include <iostream>
#include <queue>

using namespace std;

int N;
int ocean[20][20];
int start_r, start_c;
int visit[20][20];

void set_input() {
    cin >> N;
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            cin >> ocean[i][j];
            if (ocean[i][j] == 9) {
                ocean[i][j] = 0;
                start_r = i;
                start_c = j;
            }
            visit[i][j] = 0;
        }
    }
}

void printVisit() {
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            cout << visit[i][j];
        }
        cout << endl;
    }
    cout << endl;
}

void resetVisit() {
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            visit[i][j] = 0;
        }
    }
}

int bfs() {
    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    int max_time = 0;
    queue<pair<int, int> > q;
    q.push(make_pair(start_r, start_c));
    visit[start_r][start_c] = 1;

    int r, c;
    int nr, nc;
    int size = 2;
    int eat_num = 0;
    int time = 0;
    while (!q.empty()) {
        int q_size = q.size();
        pair<int, int> next_point = make_pair(-1, -1);
        time++;
        for (int i=0; i<q_size; i++) {
            r = q.front().first;
            c = q.front().second;
            q.pop();

            for (int k=0; k<4; k++) {
                nr = r + dr[k];
                nc = c + dc[k];
                if (nr<0 || nc<0 || nr>=N || nc>=N)
                    continue;
                if (visit[nr][nc] == 1 || ocean[nr][nc] > size)
                    continue;
                if (ocean[nr][nc] != 0 && ocean[nr][nc] < size) { // eat fish
                    if (next_point.first==-1 || next_point.first > nr || (next_point.first==nr && next_point.second>nc)){
                        next_point.first = nr;
                        next_point.second = nc;
                    }
                } else {
                    q.push(make_pair(nr, nc));
                    visit[nr][nc] = 1;
                }
            }
        }
        if (next_point.first != -1) {
            eat_num++;
            if (eat_num == size) {
                eat_num = 0;
                size++;
            }
            max_time = time;
            while (!q.empty()) q.pop();
            nr = next_point.first;
            nc = next_point.second;
            ocean[nr][nc] = 0;
            q.push(make_pair(nr, nc));
            visit[nr][nc] = 1;
            resetVisit();
        }
    }
    return max_time;
}

int main() {
    set_input();
    int ans = bfs();
    cout << ans << endl;
}

File: week4/16236/16236.py
from collections import deque


class P_16236:
    def __init__(self):
        self.N = int(input())
        self.ocean = []
        for i in range(self.N):
            self.ocean.append(list(map(int, input().split())))
            for j in range(self.N):
                if self.ocean[i][j] == 9:
                    self.start = (i, j)
                    self.ocean[i][j] = 0
        self.visit = [[0 for _ in range(self.N)] for _ in range(self.N)]
    
    def run(self):
        max_time = self.bfs()
        print(max_time)
    
    def bfs(self):
        max_time = 0

        dr = [-1, 0, 0, 1]
        dc = [0, -1, 1, 0]
        q = deque()

        q.append((self.start[0], self.start[1]))
        self.visit[self.start[0]][self.start[1]] = 1

        time = 0
        size = 2
        eat_num = 0
        while q:
            time += 1
            iter_size = len(q)
            eaten_fish = None
            for _ in range(iter_size):
                r, c = q.popleft()
                for k in range(4):
                    nr = r + dr[k]
                    nc = c + dc[k]
                    if nr<0 or nc<0 or nr>=self.N or nc>=self.N:
                        continue
                    if self.visit[nr][nc] == 1 or self.ocean[nr][nc] > size:
                        continue
                    if self.ocean[nr][nc] != 0 and self.ocean[nr][nc] < size: # eat fish
                        if eaten_fish is None:
                            eaten_fish = (nr, nc)
                        elif eaten_fish[0]>nr or (eaten_fish[0]==nr and eaten_fish[1]>nc):
                            eaten_fish = (nr, nc)
                        continue
                    q.append((nr, nc))
                    self.visit[nr][nc] = 1
            if eaten_fish is not None: # start from new point
                max_time = max(max_time, time)
                r, c = eaten_fish
                self.ocean[r][c] = 0
                self.visit = [[0 for _ in range(self.N)] for _ in range(self.N)]
                self.visit[r][c] = 1
                q.clear()
                q.append((r, c))
                eat_num += 1
                if eat_num == size:
                    size += 1
                    eat_num = 0
            
        return max_time

if __name__ == "__main__":
    P = P_16236()
    P.run()